api/server/router/swarm/helpers.go

```diff
@@ -34,22 +35,35 @@ func (sr *swarmRouter) swarmLogs(ctx context.Context, w http.ResponseWriter, r *
 			ShowStderr: stderr,
 			Details:    httputils.BoolValue(r, "details"),
 		},
-		OutStream: w,
+		Messages: make(chan *backend.LogMessage, 1),
 	}
 
-	chStarted := make(chan struct{})
-	if err := sr.backend.ServiceLogs(ctx, selector, logsConfig, chStarted); err != nil {
-		select {
-		case <-chStarted:
-			// The client may be expecting all of the data we're sending to
-			// be multiplexed, so send it through OutStream, which will
-			// have been set up to handle that if needed.
-			stdwriter := stdcopy.NewStdWriter(w, stdcopy.Systemerr)
-			fmt.Fprintf(stdwriter, "Error grabbing service logs: %v\n", err)
-		default:
+	tty := false
+	// checking for whether logs are TTY involves iterating over every service
+	// and task. idk if there is a better way
+	for _, service := range selector.Services {
+		s, err := sr.backend.GetService(service)
+		if err != nil {
+			// maybe should return some context with this error?
 			return err
 		}
+		// set TTY to true if the spec has a TTY OR if TTY is already true
```

Comment:
I appreciate the attention to commenting but I think we can do without this one :)